File: 415_AddStrings.cpp
//9.51%
class Solution {
public:
    string addStrings(string num1, string num2) {
		int size1(num1.size());
		int size2(num2.size());
		int maxSize(size1);
		if(size1 >= size2){
			for(int i = 0; i < size1 - size2; i++)
				num2 = "0" + num2;
		}else{
			maxSize = size1;
			for(int i = 0; i < size2 - size1; i++)
				num1 = "0" + num1;
		}
		int remain(0);
		string res;
		for(string::reverse_iterator r1 = num1.rbegin(), r2 = num2.rbegin(); r1 != num1.rend(); r1++, r2++){
			int num = (*r1 - '0') + (*r2 - '0') + remain;
			remain = num / 10;
			res = char('0' + num % 10) + res;
		}
		if(remain)
			res = '1' + res;
		return res;	
    }
};

class Solution {
public:
string addStrings(string num1, string num2) {
    int i = num1.size() - 1;
    int j = num2.size() - 1;
    int carry = 0;
    string res = "";
    while(i>=0 || j>=0 || carry){
        long sum = 0;
        if(i >= 0){sum += (num1[i] - '0');i--;}
        if(j >= 0){sum += (num2[j] - '0');j--;}
        sum += carry; 
        carry = sum / 10;
        sum = sum % 10;
        res =  res + to_string(sum);
    }
    reverse(res.begin(), res.end());
    return res;
}
};

//A better solution which use O(1) space
class Solution {
public:
    string addStrings(string num1, string num2) {
        if (num1.size() < num2.size()) return addStrings(num2, num1);
        int carry = 0, i = num1.size() - 1, j = num2.size() - 1;
        for (; i >= 0 && (carry || j >= 0); i--, j--, carry /= 10) 
            num1[i] = (carry += num1[i] - '0' + (j >= 0 ? num2[j] - '0' : 0)) % 10 + '0';
        return (carry ? "1" : "") + num1;
    }
};

File: 415_AddStrings.py
# 27.37%
class Solution(object):
    def addStrings(self, num1, num2):
        """
        :type num1: str
        :type num2: str
        :rtype: str
        """
        l1, l2 = map(int, num1), map(int, num2)
        len1, len2 = len(l1), len(l2)
        if len1 > len2:
            l2 = [0] * (len1 - len2) + l2
        else:
            l1 = [0] * (len2 - len1) + l1
        res = list()
        remain = 0
        for n1, n2 in zip(l1[::-1], l2[::-1]):
            num = (n1 + n2 + remain)
            res.insert(0, str(num % 10))
            remain = num / 10
        if remain == 1:
            res.insert(0, '1')
        return ''.join(res)
		
def addStrings(self, num1, num2):
    z = itertools.izip_longest(num1[::-1], num2[::-1], fillvalue='0')
    res, carry, zero2 = [], 0, 2*ord('0')
    for i in z:
        cur_sum = ord(i[0]) + ord(i[1]) - zero2 + carry
        res.append(str(cur_sum % 10))
        carry = cur_sum // 10
    return ('1' if carry else '') + ''.join(res[::-1])
	
from itertools import izip_longest
class Solution(object):
    def addStrings(self, num1, num2):
        res, c = "", 0
        for (x, y) in izip_longest(num1[::-1], num2[::-1], fillvalue='0'):
            s = (int(x) + int(y) + c)
            d, c = s % 10, int(s / 10)
            res = str(d) + res
        
        if c > 0: res = str(c) + res

        return res